File: services/ark-api/ark-api/src/ark_api/api/v1/namespaces.py
"""Namespaces API endpoints."""
import logging
import os
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from kubernetes_asyncio import client
from kubernetes_asyncio.client.api_client import ApiClient
from kubernetes_asyncio.client.exceptions import ApiException

from ark_sdk.models.kubernetes import NamespaceResponse, NamespaceListResponse, NamespaceCreateRequest
from ark_sdk.impersonation import ImpersonationConfig

from ...auth.dependencies import get_impersonation_config
from ...core.namespace import get_current_context
from ...models.context import ContextResponse
from .client_utils import get_impersonating_api_client
from .exceptions import handle_k8s_errors

logger = logging.getLogger(__name__)

router = APIRouter(tags=["namespaces"])


@router.get("/namespaces", response_model=NamespaceListResponse)
@handle_k8s_errors(operation="list", resource_type="namespace")
async def list_namespaces(impersonation: Optional[ImpersonationConfig] = Depends(get_impersonation_config)) -> NamespaceListResponse:
    """
    List all available namespaces.

    Returns:
        NamespaceListResponse: List of all available namespaces
    """
    async with get_impersonating_api_client(impersonation) as api:
        v1 = client.CoreV1Api(api)
        
        # List all namespaces
        namespace_list_response = await v1.list_namespace()
        
        # Convert to our response format
        namespace_list = [
            NamespaceResponse(name=ns.metadata.name)
            for ns in namespace_list_response.items
        ]
        
        return NamespaceListResponse(
            items=namespace_list,
            count=len(namespace_list)
        )


@router.post("/namespaces", response_model=NamespaceResponse)
@handle_k8s_errors(operation="create", resource_type="namespace")
async def create_namespace(body: NamespaceCreateRequest, impersonation: Optional[ImpersonationConfig] = Depends(get_impersonation_config)) -> NamespaceResponse:
    """
    Create a new Kubernetes namespace.

    Args:
        body: The namespace creation request

    Returns:
        NamespaceResponse: The created namespace details
    """
    async with get_impersonating_api_client(impersonation) as api:
        v1 = client.CoreV1Api(api)
        
        # Create the namespace object
        namespace_body = client.V1Namespace(
            metadata=client.V1ObjectMeta(name=body.name)
        )
        
        # Create the namespace
        created_namespace = await v1.create_namespace(body=namespace_body)
        
        return NamespaceResponse(
            name=created_namespace.metadata.name
        )


@router.get("/context", response_model=ContextResponse)
async def get_context_endpoint(namespace: str = None) -> ContextResponse:
    """
    Get the current Kubernetes context information.

    Returns context following standard k8s patterns:
    1. In-cluster service account (when running in pods)
    2. Kubeconfig context (when running locally)
    3. Fallback to default

    Args:
        namespace: Optional namespace to check for demo mode

    Returns:
        ContextResponse: The current namespace, cluster, and read-only mode status
    """
    current_context = get_current_context()
    
    # Use provided namespace or fall back to current context namespace
    target_namespace = namespace or current_context["namespace"]
    
    # Check if namespace exists and has demo label
    read_only_mode = False
    try:
        async with ApiClient() as api:
            v1 = client.CoreV1Api(api)
            ns = await v1.read_namespace(name=target_namespace)

            # Check if namespace has demo label
            if ns.metadata.labels and ns.metadata.labels.get("ark.mckinsey.com/demo") == "true":
                read_only_mode = True
    except ApiException as e:
        if e.status == 404:
            # Namespace doesn't exist - return 404 with default namespace for redirect
            default_namespace = current_context["namespace"]
            raise HTTPException(
                status_code=404,
                detail={
                    "message": f"Namespace '{target_namespace}' not found",
                    "default_namespace": default_namespace
                }
            )
        logger.warning("Could not check namespace labels: %s", e)
        # Fall back to environment variable for other errors
        read_only_mode = os.getenv("READ_ONLY_MODE", "false").lower() == "true"
    except Exception as e:
        logger.warning("Could not check namespace labels: %s", e)
        # Fall back to environment variable if we can't check the namespace
        read_only_mode = os.getenv("READ_ONLY_MODE", "false").lower() == "true"

    return ContextResponse(
        namespace=target_namespace,
        cluster=current_context["cluster"],
        read_only_mode=read_only_mode
    )

File: services/ark-dashboard/ark-dashboard/lib/services/namespaces.ts
import { trackEvent } from '@/lib/analytics/singleton';
import { apiClient } from '@/lib/api/client';
import type { components } from '@/lib/api/generated/types';

// Use the generated type from OpenAPI
export type Namespace = components['schemas']['NamespaceResponse'] & {
  id: number;
};
export type NamespaceListResponse =
  components['schemas']['NamespaceListResponse'];
export type NamespaceCreateRequest =
  components['schemas']['NamespaceCreateRequest'];
export type ContextResponse = components['schemas']['ContextResponse'];
export type Permissions = components['schemas']['PermissionsResponse'];

// Service with list and create operations
export const namespacesService = {
  // Get all namespaces
  async getAll(): Promise<Namespace[]> {
    const response =
      await apiClient.get<NamespaceListResponse>('/api/v1/namespaces');
    return response.items.map((item, index) => ({
      ...item,
      id: index,
    }));
  },

  // Get current Kubernetes context
  async getContext(namespace?: string): Promise<ContextResponse> {
    const response = await apiClient.get<ContextResponse>('/api/v1/context', {
      params: namespace ? { namespace } : undefined,
    });
    return response;
  },

  // Create a new namespace
  async create(name: string): Promise<Namespace> {
    const request: NamespaceCreateRequest = { name };
    const response = await apiClient.post<
      components['schemas']['NamespaceResponse']
    >('/api/v1/namespaces', request);
    trackEvent({
      name: 'namespace_created',
      properties: { namespaceName: response.name },
    });
    return {
      ...response,
      id: 0, // Will be properly assigned when we refresh the list
    };
  },
};
